client/state/utils.js

```diff
@@ -23,6 +23,84 @@ export function isValidStateWithSchema( state, schema ) {
 }
 
 /**
+ * Creates a super-reducer as a map of reducers over keyed objects
+ *
+ * Use this when wanting to write reducers that operate
+ * on a single object as if it lived in isolation when
+ * really it lives in a map of similar objects referenced
+ * by some predesignated key or id. This could be used
+ * for example when reducing properties on a site object
+ * wherein we have many sites keyed by site id.
+ *
+ * Note! This will only apply the supplied reducer to
+ * the item referenced by the supplied key in the action.
+ *
+ * If no key exists whose name matches the given keyName
+ * then this super-reducer will abort and return the
+ * previous state.
+ *
+ * If some action should apply to every single item
+ * in the map of keyed objects, this utility cannot be
+ * used as it will only reduce the referenced item.
+ *
+ * @example
+ * const age = ( state = 0, action ) =>
+ *     GROW === action.type
+ *         ? state + 1
+ *         : state
+ *
+ * const title = ( state = 'grunt', action ) =>
+ *     PROMOTION === action.type
+ *         ? action.title
+ *         : state
+ *
+ * const userReducer = combineReducers( {
+ *     age,
+ *     title,
+ * } )
+ *
+ * export default keyedReducer( 'username', userReducer )
+ *
+ * dispatch( { type: GROW, username: 'hunter02' } )
+ *
+ * state.users === {
+ *     hunter02: {
+ *         age: 1,
+ *         title: 'grunt',
+ *     }
+ * }
+ *
+ * @param {string} keyName name of key in action referencing item in state map
+ * @param {function} reducer applied to referenced item in state map
+ * @return {function} super-reducer applying reducer over map of keyed items
+ */
+export const keyedReducer = ( keyName, reducer ) => ( state = {}, action ) => {
+	// the action must refer to some item in the map
+	const itemKey = action[ keyName ];
+
+	// if no reference exists abort
+	// and return unchanged state
+	if ( ! itemKey ) {
+		return state;
+	}
+
+	// pass the old sub-state from that item into the reducer
+	const oldItemState = state[ itemKey ];
+	const newItemState = reducer( oldItemState, action );
+
+	// and do nothing if the new sub-state matches the old sub-state
+	if ( newItemState === oldItemState ) {
```

Comment:
Here's a possible fix:

```diff
diff --git a/client/state/test/utils.js b/client/state/test/utils.js
index feb4f3f..833eac9 100644
--- a/client/state/test/utils.js
+++ b/client/state/test/utils.js
@@ -281,6 +281,11 @@ describe( 'utils', () => {
                        const keyed = keyedReducer( 'name', age );
                        expect( keyed( prevState, { type: 'STAY', name: 'Bonobo' } ) ).to.equal( prevState );
                } );
+
+               it( 'should not initialize a state if no changes and not keyed', () => {
+                       const keyed = keyedReducer( 'name', age );
+                       expect( keyed( prevState, { type: 'STAY', name: 'Calypso' } ) ).to.equal( prevState );
+               } );
        } );
 
        describe( '#withSchemaValidation', () => {
diff --git a/client/state/utils.js b/client/state/utils.js
index 97ac949..91e81c9 100644
--- a/client/state/utils.js
+++ b/client/state/utils.js
@@ -85,7 +85,9 @@ export const keyedReducer = ( keyName, reducer ) => ( state = {}, action ) => {
        }
 
        // pass the old sub-state from that item into the reducer
-       const oldItemState = state[ itemKey ];
+       const oldItemState = state.hasOwnProperty( itemKey )
+               ? state[ itemKey ]
+               : reducer( undefined, { type: '@@calypso/INIT' } );
        const newItemState = reducer( oldItemState, action );
 
        // and do nothing if the new sub-state matches the old sub-state
```

Though perhaps we want it to initialize to the default value? Otherwise in the test cases if we had a `RESET` type handler which explicitly returned `0`, dispatching this action type would never assign into state a key if not already in the object.